### PR TITLE
security: use separate CA and server certs for the SQL service (separate from the RPC service)

### LINKED ISSUE #39326: Allow separate TLS certificates for SQL-only port
@knz started the work to split SQL traffic onto a separate port (#39305). Part of the value of splitting this port is that it could enable operators to use a different server certificate for SQL client communication, the same way an arbitrary server certificate can be used for the web UI today. We could also enable a different CA to validate SQL client traffic (for e.g. a Cockroach cluster that uses a Cockroach CA for all node-to-node communication but an enterprise CA for SQL clients).

After discussing the [current list of certificates](https://github.com/cockroachdb/cockroach/blob/master/pkg/security/certificate_loader.go#L74-L85) with @mberhault, we came up with the following list of additional certificates that we will need to support:

- CA cert to verify SQL clients
- CA cert that issues SQL server cert (not to be exposed or generated directly, but probably necessary for tests)
- Server cert (and key) to use for communication with SQL clients

### PR DESCRIPTION
**Pull request by knz**
Fixes #92069.
Fixes #39326.

TLDR: this sequence of commits makes it possible to use separate *server* certs for the RPC and SQL service when running CockroachDB nodes. This is possible even when both services are served from the same TCP port.

This turned to be *way* easier than expected because the TLS config paths for RPC and SQL were already cleanly separated in the code.

We're defining support for:
- optional `sql-server.crt` (and key). When this does not exist, crdb serves SQL using `node.crt` as usual.
- optional `ca-sql.crt` (and key); note that a separate CA can be used even when this file is not present by prepending it to `sql-server.crt`.
- new CLI commands `cockroach cert create-sql-server-ca` and `cockroach cert create-sql-server`.

See the individual commits for details.

**Note: this PR is not the time where we are doing major refactors/cleanups of this package. We're following established patterns here. This makes the change easily backportable.**

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM, although the naming is starting to get confusing. `ca-sql-server.crt` is used in the *client* configuration. And if we split `ca-client.crt` into SQL and RPC versions, the SQL server port would use `ca-sql-client.crt` as its CA for accepting client connections. Maybe that's OK? I can't decide whether it's confusing or not. 

In one of the test files we have a comment that `sql-server.crt` is signed by `ca-sql-server.crt`. That's helpful and should probably be in a non-test `certnames` file.

**Review by knz — COMMENTED** [review]
> the naming is starting to get confusing. ca-sql-server.crt is used in the client configuration.

I can see how that's confusing. I've shortened to `ca-sql.crt`, which follows the pattern of `ca-ui.crt` already. I think that's better.

> if we split ca-client.crt into SQL and RPC versions, the SQL server port would use ca-sql-client.crt as its CA for accepting client connections.

Yes, this will be the naming if/when we split them. But note how i've chosen not to go that route here just yet.

> That's helpful and should probably be in a non-test certnames file.

Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92126#-:-NHJrXUb6RJUcR_drzZA:bykq372)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @catj-cockroach)

**Comment by knz** [comment]
@JeffSwenson can you assist here. This PR as stated wants to split service certs between RPC and SQL _for client apps_.

In standalone SQL servers, there's currently confusion between:
- the *client* cert used by the SQL server to authn itself to the KV layer
- the *server* cert presented by the SQL server to RPC peers
- the *server* cert presented by the SQL server to SQL clients

All these currently use a single cert (confusingly called "client tenant cert").
We want to do away with this. 
I'm modifying this PR so that the server cert for SQL is using the same pattern as established here, if `sql-server.crt` exists. Do you have further thoughts?

**Comment by knz** [comment]
@JeffSwenson today the backend dial test in the sqlproxy package asserts that the server cert presented by the _SQL service_ (to the proxy) contains the tenant ID. I believe this is not necessary. Can you confirm my understanding? We only need the tenant ID when talking to the KV layer IIUC.

**Comment by jaylim-crl** [comment]
> @JeffSwenson today the backend dial test in the sqlproxy package asserts that the server cert presented by the _SQL service_ (to the proxy) contains the tenant ID. I believe this is not necessary. Can you confirm my understanding? We only need the tenant ID when talking to the KV layer IIUC.

We needed that to ensure that the proxy is routing connections to the right SQL pod. In the past, there have been cases where IP addresses were reused quickly, leading to stale routing tables within the proxy that it doesn't know about. We ended up routing a request for tenant-X to tenant-Y's SQL pod. Validating the tenant ID during routing allows the proxy to refresh its cache whenever that happens.

**Comment by knz** [comment]
Thanks. This check does not require using the tenant client cert. It merely requires including the tenant ID in the (new) SQL server cert. I'll cook up a change for that.

**Comment by bdarnell** [comment]
Does this also close #39326?

**Comment by knz** [comment]
> Does this also close https://github.com/cockroachdb/cockroach/issues/39326?

good catch. It does.

**Comment by knz** [comment]
@JeffSwenson @jaylim-crl could you review the last 6 commits of this PR? Thanks

**Review by jeffswenson — COMMENTED** [review]
LGTM

File: pkg/security/certnames/embedded.go [hunk 1]
```diff
@@ -34,6 +34,6 @@ const (
 
 // Embedded certificates specific to multi-tenancy testing.
 const (
-	EmbeddedTenantCACert = "ca-client-tenant.crt" // CA for client connections
-	EmbeddedTenantCAKey  = "ca-client-tenant.key" // CA for client connections
+	EmbeddedTenantKVClientCACert = "ca-client-tenant.crt" // CA for tenant connections to the KV layer
```
[the remark this answers is not in the document]
Comment (knz) [reply pkg/security/certnames/embedded.go]: Good point. Clarified in comment.

**Comment by knz** [comment]
@bdarnell @JeffSwenson if you like this PR either of you actually needs to go through the motion of using the github "review" button to approve; we now have automation that requires it.

**Comment by catj-cockroach** [comment]
I'm taking a look now but i wanted to make some preliminary asks that @knz and I discussed in our 1:1:

* Could we use `external` rather than `sql` for the certificate filename, and have `ui` fallback to `external`? We'd then also have `ca-external` for the CA certificate (and can have `ca-ui` fallback to `ca-external`). It's extremely unlikely that DNS and IP entries in the `ui` certificate would not be identical to the `external` certificate.
* Re: @bdarnell 's comment about the naming getting confusing, I'd argue we should deprecate `ui` certificates completely and have three (four with multi-tenancy) sets of certificates:
  * Node certificates (strictly for KV/single-tenant node-to-node communication)
  * Tenant Node certificates (strictly for SQL layer node-to-node communication)
  * External certificates (SQL, RPC, HTTP, any endpoint that would be accessed from outside the cluster)
  * Client certificates (strictly for incoming connections)

**Comment by knz** [comment]
> Could we use external rather than sql for the certificate filename, and have ui fallback to external? We'd then also have ca-external for the CA certificate (and can have ca-ui fallback to ca-external). It's extremely unlikely that DNS and IP entries in the ui certificate would not be identical to the external certificate.

We can do this yes, but I'd rather have this as an _additional_ PR on top of this one. I believe that we desire to _also_ provide the capability to use separate cert/CA pairs for http vs sql, which this PR achieves. The proposal you make would be a substitute/simpler mechanism we can add on top of it.

(In the end we want both - i'm just concerned about making this PR longer than it is already)

>  I'd argue we should deprecate ui certificates completely and have three (four with multi-tenancy) sets of certificates

We should deprecate them _in our recommended deployment docs_  (where we'd recommend using  just 1 external cert for all services) but I think we want to keep the _option_ to use separate certs for the various types of services.

As to whether the specific word "ui" is what we want to keep (vs e.g. "http") there's another debate, which will involve things around backward compat and whatnot, we we can argue this later.

**Comment by catj-cockroach** [comment]
I'd argue that we don't need to implement a difference between the `sql` certificate and the `external` certificate, given that the SQL endpoint is the one that cluster operators most likely want to set the certificate for. We may have situations where the RPC certificate and/or the HTTP certificate differ from the SQL endpoint, though I suspect that will be extremely rare. Having `external` mean "the SQL, HTTP, and RPC externally facing certificate" makes our TLS stack much easier to understand (and easier to automate with LetsEncrypt or other tools). Plus we then only need one function in `cockroach cert` for generating external certificates (as customers who intend on setting up a more complicated PKI won't use `cockroach cert` anyways)

**Comment by catj-cockroach** [comment]
*[`pkg/cli/cert.go` line 146 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/92126#-NHV_Hid8SkIavJHY2G4:-NHVVdxh2JpS3a2BuW8H:bz33ju7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61df6f6a8f7c15babe82672ea94fcdb2e6e8686/pkg/cli/cert.go#L146)):*
> ```Go
> // A createSQLServerCert command generates a SQL server certificate and stores it
> // in the cert directory.
> var createSQLServerCertCmd = &cobra.Command{
> ```

Rather than doing this, could I suggest we do `createExternalServerCertCmd`? We will never have a customer who intends on having separate SQL and UI endpoints using `cockroach cert` for certificate management.

**Comment by JeffSwenson** [comment]
> I'd argue that we don't need to implement a difference between the `sql` certificate and the `external` certificate, given that the SQL endpoint is the one that cluster operators most likely want to set the certificate for. We may have situations where the RPC certificate and/or the HTTP certificate differ from the SQL endpoint, though I suspect that will be extremely rare. Having `external` mean "the SQL, HTTP, and RPC externally facing certificate" makes our TLS stack much easier to understand (and easier to automate with LetsEncrypt or other tools). Plus we then only need one function in `cockroach cert` for generating external certificates (as customers who intend on setting up a more complicated PKI won't use `cockroach cert` anyways)

I like this simplification. Every cluster should be configured with internal certs that describe what the component is. They may be configured with external certs that are exposed to the end user via sql/http/rpc.

> I realize this is unrelated to this PR, but @JeffSwenson is there a reason we don't use node-to-node RPC for that?

My mental model of the tenant-client cert is it is the node <-> node cert for tenant sql servers. Unfortunately the tenant-client certificate predates the crdb:// uriSans. It is its own certificate type in order to allow encoding the tenant ID in the common name. I would be in favor of adding the uriSan to node certs and deprecating the tenant client cert.

**Comment by knz** [comment]
@JeffSwenson 

> Every cluster should be configured with internal certs that describe what the component is. 

I do not understand this sentence. Please rephrase/extend.

> They may be configured with external certs that are exposed to the end user via sql/http/rpc.

Ditto

**Comment by knz** [comment]
I'll try to capture the thinking in a RFC before we finalize the work here. I believe there's 2-3 different alternative proposals in this thread and i'd like to find consensus on both a strategy and some tactics to get there.
